File: Monte_Carlo_Analysis.m
%% Data-Driven Stabilization Using Prior Knowledge on Stabilizability and Controllability
% Monte Carlo Analysis
% Date: October 20, 2025
% By: Tren M.J.T. Baltussen - Eindhoven University of Technology
% and Amir Shakouri - University of Groningen
% Contact: t.m.j.t.baltussen@tue.nl
%
% See the LICENSE file in the project root for full license information.

% This script uses YALMIP and SeDuMi.

clear
close all
clc

%% System Description - Three Tank System
% Define tank areas and flow coefficients
a1 = 1; a2 = 1; a3 = 1;
k01 = 0.1; k12 = 0.5; k23 = 0.5;
x0 = [0; 0; 0];     % Initial states

% Continuous-time state-space model
A_c = [-(k01+k12)/a1,    k12/a1,     0;
        k12/a2,          -k12/a2,    k23/a2;
        0,               0,          -k23/a3];
B_c = [0; 1/a2; 0];
C_c = [1, 0, 0];
D_c = 0;

% Create continuous-time and discretized systems
sys_c = ss(A_c, B_c, C_c, D_c);
Ts = 0.1;                       % Sampling time
sys_d = c2d(sys_c, Ts);         % Discretization
A = sys_d.A; B = sys_d.B; C = sys_d.C; D = sys_d.D;

n = size(A,1);                  % State dimension
m = size(B,2);                  % Input dimension

%% Monte Carlo Simulation for Data Informativity Analysis
T = 50;                            % Number of time steps per trial
N = 100;                           % Number of Monte Carlo runs - in paper: 500.

% Initialize result arrays
check_id     = zeros(1,N);         % System identification informativity
check_ddc    = zeros(1,N);         % Data-driven control informativity
check_ddc_pk = zeros(1,N);         % Stabilization informativity (persistent kernel)

for j = 1:N
    % Generate random initial condition and input sequence
    x0 = poissrnd(1,n,1);          % Random positive integers for initial states
    Um = poissrnd(1,m,T);          % Random non-negative excitation input
    % Alternative options:
    % x0 = randn(n,1);             % Gaussian initial states
    % Um = randn(m,T);             % Gaussian input signal

    % Generate data trajectory (X-, X+)
    [Xm, Xp] = data_gen(A, B, Um, x0);

    % Evaluate informativity for identification and control
    [check_id(j), check_ddc(j), check_ddc_pk(j)] = data_analysis(Um, Xm, Xp);
end

%% Display Statistical Results
disp(strcat('Informative for stabilization: ', num2str(sum(check_ddc_pk)/N*100), '%.'))
disp(strcat('Informative for control: ', num2str(sum(check_ddc)/N*100), '%.'))
disp(strcat('Informative for system identification: ', num2str(sum(check_id)/N*100), '%.'))

%% Visualization of Informative Ratios
figure;
clf
plot(cumsum(check_id)./cumsum(ones(size(check_id))), '-', 'LineWidth', 1.5); hold on;
plot(cumsum(check_ddc)./cumsum(ones(size(check_id))), '--', 'LineWidth', 1.5);
plot(cumsum(check_ddc_pk)./cumsum(ones(size(check_id))), '-', 'LineWidth', 1.5);
ylim([0 1]);
xlabel('Simulation Index', 'Interpreter','latex');
ylabel('Informative Ratio', 'Interpreter','latex');
legend({'Identification', 'Control', 'Stabilization'}, 'Interpreter','latex');
grid on;

File: python/monte_carlo_analysis.py
"""Monte Carlo study of data informativity using the Python toolbox."""

from __future__ import annotations

import numpy as np
import matplotlib

matplotlib.use("Agg")
from matplotlib import pyplot as plt
from scipy import signal

from data_driven_stabilization import analyze_dataset, generate_state_trajectory


def three_tank_discretized(Ts: float = 0.1):
    a1 = a2 = a3 = 1.0
    k01 = 0.1
    k12 = 0.5
    k23 = 0.5

    A_c = np.array(
        [
            [-(k01 + k12) / a1, k12 / a1, 0.0],
            [k12 / a2, -k12 / a2, k23 / a2],
            [0.0, 0.0, -k23 / a3],
        ]
    )
    B_c = np.array([[0.0], [1.0 / a2], [0.0]])
    C_c = np.array([[1.0, 0.0, 0.0]])
    D_c = np.array([[0.0]])

    sys_d = signal.cont2discrete((A_c, B_c, C_c, D_c), Ts)
    A, B, _, _, _ = sys_d
    return A, B, Ts


def main():
    A, B, Ts = three_tank_discretized()
    n = A.shape[0]
    m = B.shape[1]

    T = 50
    N = 100

    rng = np.random.default_rng(0)

    check_id = np.zeros(N)
    check_ddc = np.zeros(N)
    check_ddc_pk = np.zeros(N)

    for j in range(N):
        x0 = rng.poisson(1.0, size=n)
        U = rng.poisson(1.0, size=(m, T))
        Xm, Xp = generate_state_trajectory(A, B, U, x0)
        res_id, res_ddc, res_pk = analyze_dataset(U, Xm, Xp)
        check_id[j] = int(res_id)
        check_ddc[j] = int(res_ddc)
        check_ddc_pk[j] = int(res_pk)

    print(f"Informative for stabilization: {check_ddc_pk.mean() * 100:.1f}%")
    print(f"Informative for control: {check_ddc.mean() * 100:.1f}%")
    print(f"Informative for system identification: {check_id.mean() * 100:.1f}%")

    cumulative = np.arange(1, N + 1)
    plt.figure(figsize=(8, 4))
    plt.plot(cumulative, np.cumsum(check_id) / cumulative, label="Identification")
    plt.plot(cumulative, np.cumsum(check_ddc) / cumulative, label="Control")
    plt.plot(cumulative, np.cumsum(check_ddc_pk) / cumulative, label="Stabilization")
    plt.ylim([0, 1])
    plt.xlabel("Simulation Index")
    plt.ylabel("Informative Ratio")
    plt.grid(True)
    plt.legend()
    plt.tight_layout()
    plt.savefig("monte_carlo_informativity.png", dpi=200)
    plt.close()


if __name__ == "__main__":
    main()
